proposals/nnnn-task-locals.md

```diff
@@ -0,0 +1,867 @@
+
+### Use case: Executor configuration
+
+A frequent requirement developers have voiced is to have some control and configurability over executor details on which tasks are launched.
+
+By using task locals we have a mechanism that flows naturally with the language, and due to inheritance of values also allows to automatically set up the preferred executor for tasks which do not have a preference.
+For example, invoking such actor-independent functions `calcFoo` and `calcBar` could be scheduled on specific executors (or perhaps, allow configuring executor settings) by setting a task local value like this:
+ 
+```swift
+// Just ideas, not actual API proposal (!)
+async let foo = Task.withLocal(\.executor, boundTo: someSpecificExecutor) {
+  calcFoo()
+}
+async let bar = Task.withLocal(\.executor, boundTo: .UI) {
+  calcBar()
+}
+```
+
+## Future Directions
+
+### Access from non-async functions
+
+Today, it is not possible to access a `Task` from a non-asynchronous function. 
+
+This is troublesome because it means that refactoring any code that needs to access a task-local must be refactored into an asynchronous function, introducing a potential suspension point (even though task local value reads *never* actually suspend execution).
+
+This limitation will be lifted as the ABI of async functions is locked down and stabilized in the coming versions of Swift.
+
+One these issues are resolved, the Task API is expected to gain API similar to these:
+
+```swift
+extension Task {
+  func unsafeCurrent() -> UnsafeCurrentTask?
+}
+```
+
+Where the `unsafeCurrent` allows access to the task a function is currently executing in, even if called from a synchronous function. The `Unsafe` part of the name originates from the fact that it is NOT safe to modify such task object from _other_ tasks, and such developers must not pass such task object to other tasks. Similar semantics of "must not escape current scope" also exist for task groups (from the Structured Concurrency proposal), as such it is expected we will be able to express such `@unmoveable` restriction once move-only types arrive in the language. Then, the `Unsafe...` part of this type would no longer be necessary as we could enforce the safety of it. 
+
+Using such API, in the future, we will be able to access task locals from non-async functions like this:
+
+```swift
+func synchronous() {
+  if let task: UnsafeCurrentTask = Task.unsafeCurrent() {
+    task.local(\.name)
+  }
+}
+```
+
+The function `local(_:)` is defined on the `UnsafeCurrentTask` itself _on purpose_ as to avoid the more expensive lookup that may be involved with `unsafeCurrent()`. These APIs may change and are subject to a different proposal however.
```

Comment:
```suggestion
The function `local(_:)` is defined on the `UnsafeCurrentTask` itself _on purpose_ so as to avoid the more expensive lookup that may be involved with `unsafeCurrent()`. These APIs may change and are subject to a different proposal however.
```